### PR TITLE
Add support for outetts

### LINKED ISSUE #28: Find and implement parler_tts alternative
Currently, there is an issues installing parler_tts inside codespaces. It uses a dependency that also tries to install from git and fails. In the short-term, we've forked parler_tts to trim dependencies.
Long-term, we'd like to find an alternative to parer_tts without these dependency issues in codespaces.

**Acceptance criteria**
- Alternative to parler_tts installs in codespaces without issues 
- Alternative to parler_tts gives comparable results in terms of speech quality
- - Alternative to parler_tts works in similar speed 
- Alternative can be implemented and works as part of end-to-end Blueprint

**UAT**
As a user, when I upload my document, I can generate a natural sounding audio podcast.

### PR DESCRIPTION
**Pull request by Kostis-S-Z**
# What's changing
- Add support for Oute TTS model family
- Change default TTS model in demo app and tests from Parler to Oute
- Make Parler optional dependency 
- Update `.github/workflows/tests.yaml` to also install the `parler` dependency

Closes #28 

# How to test it

Steps to test the changes:

1. Install package `pip install -e .`
2. Run demo `python -m streamlit run demo/app.py`
3. Generate a podcast
4. :tada: 

# Additional notes for reviewers

- If the user tries to change in the `demo/app.py` the `load_text_to_speech_model` function to use a `parler` model, the app will crash. Should we add support for the `parler` models in the demo app? (it just adds more model-specific code which is not great for maintainability)

- We installing the _optional_ parler dependency, we went back to installing the original parler package instead of @daavoo 's fork (`"parler_tts @ git+https://github.com/huggingface/parler-tts.git"`)

- I feel like the docs could use some extra love (they are not outdated, but they also dont reflect the full potential / functionality). Maybe needs a follow-up PR?

# I already...

- [x] Tested the changes in a working environment to ensure they work as expected
- [x] Added some tests for any new functionality
- [x] Updated the documentation (both comments in code and under `/docs`)

### CODE REVIEW

**Comment by Kostis-S-Z** [comment]
@stefanfrench from this PR quite a few of the docs got outdated. I started an effort to update them but I would highly appreciate your help if you have time! I took the liberty on adding / modifying some content, please take a look and adjust it any way you see fit!

File: demo/app.py [hunk 1]
```diff
@@ -86,59 +131,64 @@ def load_text_to_speech_model_and_tokenizer():
...
 
-    st.markdown(
-        "For this demo, we are using the following models: \n"
-        "- [OLMoE-1B-7B-0924-Instruct-GGUF](https://huggingface.co/allenai/OLMoE-1B-7B-0924-Instruct-GGUF)\n"
-        "- [parler-tts-mini-v1](https://huggingface.co/parler-tts/parler-tts-mini-v1)"
-    )
-    st.markdown(
-        "You can check the [Customization Guide](https://mozilla-ai.github.io/document-to-podcast/customization/)"
-        " for more information on how to use different models."
+    model_name = st.selectbox(
```
Comment (daavoo) [inline demo/app.py:136]: @Kostis-S-Z Unless I misunderstood, I thought we agree on picking 1 model for the demo and document how to use a different one in the customization guide (using the CLI and/or direct API usage)

**Review by daavoo — COMMENTED** [review]
@Kostis-S-Z In #34 I am changing things for the CLI usage (i.e. `config` is used differently, `podcast_maker` is not used at all, and tests have changed).
We can either merge this first and I will update everything on the CLI PR or wait for the CLI PR and update here.

File: demo/app.py (hunk 1 above)
Comment (Kostis-S-Z) [reply demo/app.py:136]: Indeed thats what we said, and I dont mind re-implementing it like that. I just thought that this is more user-friendly and doesnt cost us anything :man_shrugging: . I ll implement whatever we think its better for UX

**Comment by Kostis-S-Z** [comment]
> @Kostis-S-Z In #34 I am changing things for the CLI usage (i.e. `config` is used differently, `podcast_maker` is not used at all, and tests have changed). We can either merge this first and I will update everything on the CLI PR or wait for the CLI PR and update here.

I see. Is the CLI PR almost done? If so then lets merge that first and then I ll update.

File: demo/app.py (hunk 1 above)
Comment (stefanfrench) [reply demo/app.py:136]: @Kostis-S-Z  Lets go ahead with just picking 1 model for the demo - makes its easier for it to be robust with 1 model + prompt combo. @daavoo 's CLI branch should make it easier for people to switch if they want to.

**Comment by daavoo** [comment]
@Kostis-S-Z I pushed some changes because there were some conflicts with `main` + fixed some tests

**Comment by Kostis-S-Z** [comment]
> @Kostis-S-Z I pushed some changes because there were some conflicts with `main` + fixed some tests

@daavoo awesome, thanks for helping out!

File: docs/step-by-step-guide.md [hunk 2]
```diff
@@ -79,21 +77,29 @@ In this final step, the generated podcast transcript is brought to life as an au
...
+   - The function `text_to_speech` takes the input text (e.g. podcast script) and speaker profile, generating a waveform (audio data in a numpy array) that represents the spoken version of the text.
 
 **2 - Parsing and Combining Voices**
 
 - The [`script_to_audio.py`](api.md/#document_to_podcast.podcast_maker.script_to_audio) script ensures each speaker’s dialogue is spoken in their unique voice.
 
 - The function `parse_script_to_waveform` splits the dialogue script by speakers and uses `text_to_speech` to generate audio for each speaker, stitching them together into a full podcast.
 
-- Once the podcast waveform is ready, the save_waveform_as_file function saves it as an audio file (e.g., MP3 or WAV), making it ready for distribution.
+- Once the podcast waveform is ready, the `save_waveform_as_file` function saves it as an audio file (e.g., MP3 or WAV), making it ready for distribution.
```
Comment (daavoo) [inline docs/step-by-step-guide.md:102]:
```suggestion
- Once the podcast waveform is ready, the `soundfile.write` function saves it as an audio file (e.g., MP3 or WAV), making it ready for distribution.
```

File: example_data/config.yaml [hunk 3]
```diff
@@ -22,8 +22,9 @@ speakers:
...
     name: Laura
     description: The main host. She explains topics clearly using anecdotes and analogies, teaching in an engaging and captivating way.
-    voice_profile: Laura's voice is exciting and fast in delivery with very clear audio and no background noise.
+    voice_profile: female_1
   - id: 2
     name: Jon
     description: The co-host. He keeps the conversation on track, asks curious follow-up questions, and reacts with excitement or confusion, often using interjections like hmm or umm.
-    voice_profile: Jon's voice is calm with very clear audio and no background noise.
+    voice_profile: male_1
+device: "cpu"
```
Comment (daavoo) [inline example_data/config.yaml:30]: I think it doesn't make sense to have this as a parameter. 
We haven't tested gpu support yet and once we have it, I think is better to use `torch.cuda.is_available` to set the device automatically.

```suggestion
```

File: src/document_to_podcast/inference/text_to_speech.py [hunk 4]
```diff
@@ -1,27 +1,72 @@
+from typing import Union
+
 import numpy as np
-from transformers import PreTrainedModel, PreTrainedTokenizerBase
+from outetts.version.v1.interface import InterfaceGGUF
+from transformers import PreTrainedTokenizerBase, PreTrainedModel
 
 
-def text_to_speech(
+def _speech_generation_oute(
```
Comment (daavoo) [inline src/document_to_podcast/inference/text_to_speech.py:8]: No strong opinion but why not use the same name as the main function:

```suggestion
def _text_to_speech_oute(
```

File: src/document_to_podcast/inference/text_to_speech.py [hunk 5]
```diff
@@ -1,27 +1,72 @@
...
+        repetition_penalty=1.1,
+        max_length=4096,
+        speaker=speaker,
+    )
+
+    output_as_np = output.audio.cpu().detach().numpy().squeeze()
+    return output_as_np
+
+
+def _speech_generation_parler(
```
Comment (daavoo) [inline src/document_to_podcast/inference/text_to_speech.py:28]:
```suggestion
def _text_to_speech_parler(
```

**Review by daavoo — COMMENTED** [review]
Left some minor comments. Will now test on codespaces

File: demo/app.py [hunk 6]
```diff
@@ -23,8 +23,8 @@ def load_text_to_text_model():
 
 
 @st.cache_resource
-def load_text_to_speech_model_and_tokenizer():
-    return load_parler_tts_model_and_tokenizer("parler-tts/parler-tts-mini-v1", "cpu")
+def load_text_to_speech_model():
+    return load_outetts_model("OuteAI/OuteTTS-0.1-350M-GGUF/OuteTTS-0.1-350M-FP16.gguf")
```
Comment (daavoo) [inline demo/app.py:27]: I think we should use this one. It is much better quality. The 0.1-350 collapses similar to the parler_tts.

```suggestion
    return load_outetts_model("OuteAI/OuteTTS-0.2-500M-GGUF/OuteTTS-0.2-500M-FP16.gguf")
```

**Review by daavoo — CHANGES_REQUESTED** [review]

**Review by stefanfrench — APPROVED** [review]
@Kostis-S-Z - Tested this locally from my side and it worked well end-to-end. 

350M TTS outputs are quite poor, so lets proceed with using 500M as the default TTS model for now.

File: src/document_to_podcast/cli.py [hunk 7]
```diff
@@ -30,9 +30,10 @@ def document_to_podcast(
     output_folder: str | None = None,
     text_to_text_model: str = "allenai/OLMoE-1B-7B-0924-Instruct-GGUF/olmoe-1b-7b-0924-instruct-q8_0.gguf",
     text_to_text_prompt: str = DEFAULT_PROMPT,
-    text_to_speech_model: SUPPORTED_TTS_MODELS = "parler-tts/parler-tts-mini-v1",
+    text_to_speech_model: SUPPORTED_TTS_MODELS = "OuteAI/OuteTTS-0.1-350M-GGUF/OuteTTS-0.1-350M-FP16.gguf",
     speakers: list[Speaker] | None = None,
     from_config: str | None = None,
+    device: str = "cpu",
```
Comment (daavoo) [inline src/document_to_podcast/cli.py:36]:
```suggestion
```

File: src/document_to_podcast/cli.py [hunk 8]
```diff
@@ -65,13 +66,15 @@ def document_to_podcast(
...
         text_to_speech_model (str, optional): The path to the text-to-speech model.
-            Defaults to `parler-tts/parler-tts-mini-v1`.
+            Defaults to `OuteAI/OuteTTS-0.1-350M-GGUF/OuteTTS-0.1-350M-FP16.gguf`.
 
         speakers (list[Speaker] | None, optional): The speakers for the podcast.
             Defaults to DEFAULT_SPEAKERS.
 
         from_config (str, optional): The path to the config file. Defaults to None.
 
+        device (str, optional): The device to load the model. Currently ONLY the TTS! Defaults to "cpu".
```
Comment (daavoo) [inline src/document_to_podcast/cli.py:76]:
```suggestion
```

File: src/document_to_podcast/cli.py [hunk 9]
```diff
@@ -85,6 +88,7 @@ def document_to_podcast(
             text_to_text_prompt=text_to_text_prompt,
             text_to_speech_model=text_to_speech_model,
             speakers=[Speaker.model_validate(speaker) for speaker in speakers],
+            device=device,
```
Comment (daavoo) [inline src/document_to_podcast/cli.py:91]:
```suggestion
```

File: src/document_to_podcast/inference/text_to_speech.py [hunk 10]
```diff
@@ -1,27 +1,72 @@
...
-    input_ids = tokenizer(voice_profile, return_tensors="pt").input_ids
-    prompt_input_ids = tokenizer(input_text, return_tensors="pt").input_ids
-
-    generation = model.generate(input_ids=input_ids, prompt_input_ids=prompt_input_ids)
-    waveform = generation.cpu().numpy().squeeze()
-    return waveform
+    if isinstance(model, InterfaceGGUF):
+        return _speech_generation_oute(input_text, model, voice_profile)
+    elif isinstance(model, PreTrainedModel):
+        return _speech_generation_parler(input_text, model, tokenizer, voice_profile)
```
Comment (daavoo) [inline src/document_to_podcast/inference/text_to_speech.py:70]:
```suggestion
        return _text_to_speech_parler(input_text, model, tokenizer, voice_profile)
```

**Pull request merged**